app/actions/views/custom_status.ts

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Client4} from '@client/rest';
+import {logError} from '@mm-redux/actions/errors';
+import {UserTypes} from '@mm-redux/action_types';
+import {getCurrentUser} from '@mm-redux/selectors/entities/common';
+import {ActionFunc, DispatchFunc, batchActions, GetStateFunc} from '@mm-redux/types/actions';
+import {UserCustomStatus} from '@mm-redux/types/users';
+
+export function setCustomStatus(customStatus: UserCustomStatus): ActionFunc {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const user = getCurrentUser(getState());
+        const oldCustomStatus = user.props?.customStatus;
+
+        if (!user.props) {
+            user.props = {};
+        }
+
+        user.props.customStatus = JSON.stringify(customStatus);
+        dispatch({type: UserTypes.RECEIVED_ME, data: user});
+
+        try {
+            await Client4.updateCustomStatus(customStatus);
```

Comment:
The purpose of dispatching the `RECEIVED_ME` action here was to minimize the delay that was observed while setting the status. After setting the status, there was a 1sec delay in seeing the status in the sidebar and other places because we were relying on the WebSocket events. If we dispatch the `RECEIVED_ME` action after waiting for the API response, we shall see the same delay.

Reply:
Yeah this is called an optimistic action ;)